staging/kos/cmd/attachment-tput-driver/schedule.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"math/rand"
+	"os"
+	"time"
+
+	"github.com/golang/glog"
+)
+
+// A threadOpsSchedule is the schedule of operations on attachments for a single
+// thread. It consists of a list `dt0,..,dtn` of time intervals in nanoseconds.
+// The ith operation that the thread must perform must happen at `t0 + dti`,
+// where t0 is an arbitrary start time for operations on attachments.
+// "Arbitrary start time" means that the start time is not hardcoded in the
+// schedule, and it is up to the schedule user to provide it. This is done
+// because computing a schedule might take time; if the schedule was dependent
+// on a t0, by the time it was computed t0 + dti might already have elapsed.
+// Compute the schedule first, and then pick a t0.
+type threadOpsSchedule []int64
+
+// Supported distributions of the operations on attachments.
+const (
+	steadyDistribution  = "steady"
+	poissonDistribution = "poisson"
+)
+
+// newOpsSchedule returns a schedule for the operations on attachments. The
+// returned schedule satisfies the `opsDistribution` and `opsPeriod` parameters.
+func newOpsSchedule(opsDistribution string, opsPeriod float64, totalOps, numThreads uint64) map[uint64]threadOpsSchedule {
+	threadIDToOpsSchedule := make(map[uint64]threadOpsSchedule, numThreads)
+
+	// For each operation on attachments, compute a dt such that the operation
+	// should take place at `t0 + dt` where t0 is the arbitrary start time for
+	// operations managed outside of this function.
+	var dtNanos int64
+	for i := uint64(0); i < totalOps; i++ {
+		if opsDistribution == poissonDistribution {
+			// The time between an op and the next one is given by the
+			// exponential distribution with rate `1/opsPeriod`.
+			dtNanos += int64(float64(time.Second) * (opsPeriod * rand.ExpFloat64()))
+		} else {
+			// The ops on attachments happen with a constant period.
+			dtNanos += int64(float64(time.Second) * opsPeriod)
+		}
+		if dtNanos < 0 {
+			// Overflow! This should never happen, because at every iteration
```

Comment:
This can also happen because a true exponential distribution includes some very large numbers.
I think it would be fair to truncate the exponential distribution --- discard values larger then 1000, for example.

Reply:
In the steady case, this overflow can not happen unless opsPeriod is crazy (wildly unusably) long; with a usable opsPeriod, we will run out of memory rather than hit this overflow.

Reply:
49677ba truncates the exponential distribution to 1000secs.
I would keep the check on overflows anyway, on the basis that a user could erroneously specify a big value for the period; having the check produces an informative error message.